File: src/main/java/am/adrian/dungeonkeeper/model/gameobject/character/Goblin.java
package am.adrian.dungeonkeeper.model.gameobject.character;

import am.adrian.dungeonkeeper.constant.Direction;
import am.adrian.dungeonkeeper.model.coords.MutableCoords;
import am.adrian.dungeonkeeper.model.gameobject.Creature;
import am.adrian.dungeonkeeper.model.gameobject.GameObject;
import am.adrian.dungeonkeeper.model.gameobject.trait.*;
import am.adrian.dungeonkeeper.model.health.Health;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

@RequiredArgsConstructor
@Getter
@Setter
public class Goblin implements Creature, HasHealth, CanWalk, CanSwim, HasEmotion, CanLevel {

    private static final String TEXTURE = "goblin.png";

    private final MutableCoords coords = new MutableCoords();
    private final Health health;
    private final int width;
    private final int height;

    private int level = 1;
    private Mood mood = Mood.HAPPY;

    @Override
    public String getTexture() {
        return TEXTURE;
    }

    @Override
    public int getWidth() {
        return width;
    }

    @Override
    public int getHeight() {
        return height;
    }

    @Override
    public void walk(Direction dir) {
        move(dir, walkDelta());
    }

    @Override
    public int walkDelta() {
        return 5;
    }

    @Override
    public void swim(Direction dir) {
        move(dir, swimDelta());
    }

    @Override
    public int swimDelta() {
        return 1;
    }

    @Override
    public void levelUp() {
        ++level;
    }

    public void move(Direction dir, int delta) {
        switch (dir) {
            case UP -> coords.incY(-delta);
            case DOWN -> coords.incY(delta);
            case LEFT -> coords.incX(-delta);
            case RIGHT -> coords.incX(delta);
            default -> throw new RuntimeException("Unknown direction");
        }
    }

    @Override
    public boolean collides(@NotNull GameObject gameObject) {
        return false;
    }
}

File: src/main/java/am/adrian/dungeonkeeper/model/gameobject/character/Goblin.kt
package am.adrian.dungeonkeeper.model.gameobject.character

import am.adrian.dungeonkeeper.constant.Direction
import am.adrian.dungeonkeeper.model.coords.MutableCoords
import am.adrian.dungeonkeeper.model.gameobject.GameObject
import am.adrian.dungeonkeeper.model.gameobject.trait.CanFear
import am.adrian.dungeonkeeper.model.gameobject.trait.CanLevel
import am.adrian.dungeonkeeper.model.gameobject.trait.CanSwim
import am.adrian.dungeonkeeper.model.gameobject.trait.CanWalk
import am.adrian.dungeonkeeper.model.gameobject.trait.HasHealth
import am.adrian.dungeonkeeper.model.gameobject.trait.HasMood
import am.adrian.dungeonkeeper.model.gameobject.trait.HasMood.Mood
import am.adrian.dungeonkeeper.model.health.Health

private const val TEXTURE = "goblin.png"

class Goblin(
    private val health: Health,
    private val width: Int,
    private val height: Int
) : Creature, HasHealth, CanLevel, CanWalk, CanSwim, HasMood, CanFear {

    private val coords: MutableCoords = MutableCoords()

    private var level: Int = 1
    private var mood = Mood.HAPPY

    override fun collides(gameObject: GameObject): Boolean = false

    override fun getCoords(): MutableCoords = coords

    override fun getTexture(): String = TEXTURE

    override fun getLevel(): Int = level

    override fun levelUp() {
        ++level
    }

    override fun getWidth(): Int = width

    override fun getHeight(): Int = height

    override fun walk(dir: Direction) = move(dir, walkDelta())

    override fun walkDelta(): Int = 5

    override fun swim(dir: Direction) = move(dir, swimDelta())

    override fun swimDelta(): Int = 1

    override fun getFearThreshold(): Int = getLevel() * 2

    override fun move(dir: Direction, delta: Int) {
        when (dir) {
            Direction.UP -> coords.incY(-delta)
            Direction.DOWN -> coords.incY(delta)
            Direction.LEFT -> coords.incX(-delta)
            Direction.RIGHT -> coords.incX(delta)
        }
    }

    override fun getHealth(): Health = health

    override fun getMood(): Mood = mood

    override fun setMood(mood: Mood) {
        this.mood = mood
    }
}
